cordformation/src/main/kotlin/net/corda/plugins/Dockerform.kt

```diff
@@ -146,33 +143,75 @@ open class Dockerform @Inject constructor(objects: ObjectFactory) : Baseform(obj
 
                 it.installDefaultDatabaseConfig(dbConfig)
 
-                val args = mutableMapOf<String, Any>()
-
-                dockerfileArgs.entrySet().toList().map {
-                    args[it.key] = it.value.unwrapped()
-                }
-
-                services[dbHost] = mapOf(
+                val dbDockerfile = dockerConfig.getString("dockerConfig.dbDockerfile")
+                // Override the port and hostname parameters in dockerfile
+                val dbDockerfileArgs = defaultUrlArgs.withFallback(
+                        dockerConfig.getConfig("dockerConfig.dbDockerfileArgs")).resolve()
+                val database = mutableMapOf(
                         "build" to mapOf(
                              "context" to project.buildDir.absolutePath,
                              "dockerfile" to project.buildDir.resolve(dbDockerfile).toString(),
-                             "args" to args
+                             "args" to dbDockerfileArgs.entrySet().associate { it.key to it.value.unwrapped() }
                         ),
-                        "restart" to "unless-stopped",
-                        "ports" to listOf(dbPort)
+                        "restart" to "unless-stopped"
                 )
 
+                // append persistence volume if it is required
+                if (dockerConfig.hasPath("dockerConfig.dbDataVolume")) {
+
+                    var hostPathStr = dockerConfig.getString("dockerConfig.dbDataVolume.hostPath")
+
+                    if (dockerConfig.hasPath("dockerConfig.dbDataVolume.hostPathArgs")) {
+                        val hostPathArgs = dockerConfig.getConfig("dockerConfig.dbDataVolume.hostPathArgs")
+                        hostPathStr = TypesafeUtils.resolveString(hostPathStr, hostPathArgs)
+                    }
+
+                    var containerPathStr = dockerConfig.getString("dockerConfig.dbDataVolume.containerPath")
+
+                    if (dockerConfig.hasPath("dockerConfig.dbDataVolume.containerPathArgs")) {
+                        val containerPathArgs = dockerConfig.getConfig("dockerConfig.dbDataVolume.containerPathArgs")
+                        containerPathStr = TypesafeUtils.resolveString(containerPathStr, containerPathArgs)
+                    }
+
+                    val hostPath = Paths.get(hostPathStr)
+                    var absoluteHostPath = when {
+                        hostPath.isAbsolute -> hostPath
+                        else -> nodeBuildPath.resolve(hostPath).toAbsolutePath()
+                    }
+                    val hostDir = File(absoluteHostPath.toUri())
```

Comment:
And this can still be
```kotlin
val hostDir = absoluteHostPath.toFile()
```
which means you don't need to import `java.io.File` any more.